### ISSUE TITLE
YOLOv5 Different inference result between tensorrt and pytorch

### ISSUE THREAD

**Issue by hxk11111** [wang-xinyu/tensorrtx#100]
I tried to convert yolov5s trained by my custom dataset from pytorch to tensorrt.

But when I run validation process using the same dataset, the result produced by tensorrt had a mAP more than 10 percent lower than original pytorch model.

Is that normal? Has anyone compared the result before and after converting tensorrt?

**Comment by wang-xinyu** [comment]
Please check the params,  e.g. conf, nms thresh, max output, etc.  In yololayer.h and yolov5s.cpp

**Comment by hxk11111** [comment]
> Please check the params, e.g. conf, nms thresh, max output, etc. In yololayer.h and yolov5s.cpp

I have checked all the params. Theoretically, the result should be the same, right?

**Comment by wang-xinyu** [comment]
Yes, it should be same.

**Comment by mythke** [comment]
I have the same question in Yolov5 2.0 version. Do you solve it? And I try to pytorch -> onnx -> tensorrt it also not work... So sad...

**Comment by wang-xinyu** [comment]
@mythke what the mAP you got with this repo?  COCO or your own dataset?

**Comment by mfruhner** [comment]
Now that I got it running, i also notice that the trt engine for v5 m detects way less BBoxes than the detect.py version. I lowered the thresh values by a lot already...

**Comment by wang-xinyu** [comment]
@mfruhner did you lower the `IGNORE_THRESH` in yololayer.h?

**Comment by mythke** [comment]
I find the reason.
yolov5.cpp 553 line to normlization should divide 255.0 not 255.
`data[b * 3 * INPUT_H * INPUT_W + i] = uc_pixel[2] / 255.0;`
after do this, I test in coco val 2017 mAP is 33.7, but pytorch version has 35.5. 
I guess the yolo layer parameter may not consistent

**Comment by wang-xinyu** [comment]
@mythke this should be a problem.   however, some compiler may not have this issue. I will fix it.

can you try lower the `IGNORE_THRESH` in yololayer.h, like 0.001?

**Comment by mythke** [comment]
@wang-xinyu it not work. using `IGNORE_THRESH` as 0.001. also 2 points lower than pytorch version

**Comment by mfruhner** [comment]
I lowered the IGNORE_THRESH to 0.001 and NMS and CONF to 0.1. This gives me better results, but still many objects missing, exspecially the small ones.

**Comment by wang-xinyu** [comment]
@mythke the nms method may be different in pytorch.   And also which yolo layer param do you think might impact the result?

**Comment by wang-xinyu** [comment]
@mfruhner I already fixed the bug @mythke just mentioned, can you update and try again?

**Comment by mfruhner** [comment]
Yes I tried again.. is there some sort of min box size ? Only the small objects are not recognised now

**Comment by wang-xinyu** [comment]
@mfruhner there is no box size limitations. Are you using the same input shape as pytorch?

**Comment by wang-xinyu** [comment]
@mythke I just fixed one line of the nms(), https://github.com/wang-xinyu/tensorrtx/commit/7cd092d38289123442157cf7defab78e816f4440, can you set `MAX_OUTPUT_BBOX_COUNT` to 99999999 and try again?

**Comment by mfruhner** [comment]
Thanks for the quick fixes. Increasing the MAX_OUTPUT_BBOX_COUNT did the trick. So this is before nms, right?

**Comment by wang-xinyu** [comment]
`MAX_OUTPUT_BBOX_COUNT` limits the number of output boxes from yololayer. So that the nms process will be much faster.
Usually we don't have such many objects to detect. So I set it to 1000 by default.

**Comment by mythke** [comment]
@wang-xinyu I try `MAX_OUTPUT_BBOX_COUNT` to 9999999, it also not work...

**Comment by wang-xinyu** [comment]
@mythke There is one thing, did you notice?  The input preprocessing method is `letterbox`, which has padding.   So did you minus the padding?

**Comment by mythke** [comment]
@wang-xinyu I notice this. I feed the same image processed by letterbox, BGRtoRGB and Normlization to pytorch and trt model. But the result has little difference.

**Comment by mythke** [comment]
@wang-xinyu I find some things. I guess if the predict bbox is out of image, the process maybe different. I'm watching the code. it just my guess.

**Comment by mythke** [comment]
@wang-xinyu I  feed torch model and trt model as the same input. The value of yolo layer input has a little difference. I think there are something different implement between the backbone.

**Comment by wang-xinyu** [comment]
@mythke The backbone should be the same.  But tensorrt layers may have slight difference from pytorch. 

I still think the nms may impact the result...

**Comment by mythke** [comment]
@wang-xinyu I check the backbone. It is correct. I find the pytorch model conv results has a little difference between trt results. If using a deep backbone, the difference will be large. I will seach the reason in tensorrt community. Thank you for your pretty good job!

**Comment by wang-xinyu** [comment]
@mythke Yes that should be the reason.   Thanks for your feedback.

**Comment by sudapure** [comment]
@mythke , have you figured out any reason for the difference of mAP results afterwards.
